notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,304 @@
+  |> Axon.dense(784, activation: :sigmoid, name: "decoder_layer_4")
+  # Turn it back into a 28x28 single channel image
+  |> Axon.reshape({1, 28, 28})
+
+# We can use Axon.Display to show us what each of the layers would look like
+# assuming we send in a batch of 4 images
+Axon.Display.as_table(model, Nx.template({4, 1, 28, 28}, :f32)) |> IO.puts()
+```
+
+```elixir
+batch_size = 128
+
+train_data = Utils.prepare_training_data(train_images, 128)
+test_data = Utils.prepare_training_data(test_images, 128)
+
+{input_batch, target_batch} = Enum.at(train_data, 0)
+Kino.render(input_batch[[images: 0]] |> Utils.image_to_kino())
+Kino.render(target_batch[[images: 0]] |> Utils.image_to_kino())
+
+:ok
+```
+
+When training, it can be useful to stop execution early - either when you see it's failing and you don't want to waste time waiting for the remaining epochs to finish, or if it's good enough and you want to start experimenting with it.
+
+The `livebook_early_stop/1` function below is a handy handler to give us a `Kino.Control.button` that will stop the training loop when clicked.
+
+```elixir
+defmodule LivebookTraining do
+  def livebook_early_stop(loop) do
+    frame = Kino.Frame.new() |> Kino.render()
+    stop_button = Kino.Control.button("stop")
+    Kino.Frame.render(frame, stop_button)
+
+    {:ok, button_agent} = Agent.start_link(fn -> nil end)
+
+    Task.start_link(fn ->
+      Kino.Control.subscribe(stop_button, :stop)
+
+      receive do
+        {:stop, _info} ->
+          Kino.Control.unsubscribe(stop_button)
+          Agent.update(button_agent, fn _ -> :stop end)
+      end
+    end)
+
+    handler = fn state ->
+      stop_state = Agent.get(button_agent, & &1)
+
+      if stop_state == :stop do
+        Agent.stop(button_agent)
+        Kino.Frame.render(frame, "stopped")
+        {:halt_loop, state}
+      else
+        {:continue, state}
+      end
+    end
+
+    Axon.Loop.handle(loop, :iteration_completed, handler)
+  end
+end
+```
+
+```elixir
+# A helper function to display the input and output side by side
+combined_input_output = fn params, image_index ->
+  test_image = test_images[[images: image_index]]
+  reconstructed_image = Axon.predict(model, params, test_image) |> Nx.squeeze(axes: [0])
+  Nx.concatenate([test_image, reconstructed_image], axis: :width)
+end
+
+frame = Kino.Frame.new() |> Kino.render()
+
+render_example_handler = fn state ->
+  Kino.Frame.append(frame, "Epoch: #{state.epoch}, Iteration: #{state.iteration}")
+  # state.step_state[:model_state] contains the model params when this event is fired
+  params = state.step_state[:model_state]
+  image_index = Enum.random(0..(Nx.axis_size(test_images, :images) - 1))
+  image = combined_input_output.(params, image_index) |> Utils.image_to_kino(200, 400)
+  Kino.Frame.append(frame, image)
+  {:continue, state}
+end
+
+params =
+  model
+  |> Axon.Loop.trainer(:mean_squared_error, Axon.Optimizers.adam(0.001))
```

Comment:
This is the loop that NaNs? I can't see any obvious reason why. A few things you can try:

1. Play with the `eps` option in `adam`...it's a numerical stability term, I don't think it should matter but it might here
2. Lower the learning rate and see if it still learns without collapsing to NaN. Though 1e-3 still seems small enough for this kind of problem.
3. Try BCE loss. I've seen it used successfully in Autoencoders before, but again I don't see why this would lead to NaNs

For gradient clipping, there are a few options. If you take a look at Axon's optimizers, you'll see they're all actually rather simple. For example, I think `adam` is actually just:

```elixir
Axon.Updates.scale_by_adam()
|> Axon.Updates.scale()
```

You can implement a custom one and add one of the clipping functions:

```elixir
Axon.Updates.clip_by_norm()
|> Axon.Updates.scale_by_adam()
|> Axon.Updates.scale()
```

Or you can compose existing optimizers with updates like this:

```elixir
optimizer = Axon.Updates.compose(Axon.Updates.clip_by_norm(), Axon.Optimizers.adam(1.0e-3))
```

To inspect the gradients of each layer, you can attach backward hooks to the model with `Axon.attach_hook/3`. If you want to apply one to every layer of the model without needing to do so manually, you can try:

```elixir
instrumented_model = Axon.map_nodes(model, fn layer ->
  Axon.attach_hook(layer, &IO.inspect/1, on: :backward)
end)
```

For something more robust you can probably also visualize the gradients and the results of the forward pass of intermediate layers too by attaching to the `on: :forward` event